### PR TITLE
feat(client): add json client interface

### LINKED ISSUE #425: Add JSON-API mode to `walrus-client`
Add a JSON-API mode for the client (i.e., run walrus-client json '{"some": "json input"}'), which offers the same functionality as the HTTP API.

Related to #414.

### PR DESCRIPTION
**Pull request by giac-mysten**
- [x] cli input as json
- [x] cli output as json
- [x] blobs as base64 in json
~update all cli examples~
~update all calls to cli in scripts~

Closes #425

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -92,6 +92,42 @@ curl -X PUT "http://$ADDRESS/v1/store?epochs=5" -d @"some/file" # store file `so
 curl "http://$ADDRESS/v1/<some blob ID>" # read a blob from Walrus (with aggregator or daemon)
 ```
 
+### Json mode
+
+All Walrus client commands (except, currently, the `info` command) are available in `json` mode.
+In this mode, all the command line flags of the original cli command can be specified in json format.
+The `json` mode therefore simplifies programmatic access to the cli.
```
Comment (mlegner) [inline README.md]: Nits: I would write "JSON" and "CLI" unless in code mode (`json`, `cli`). This also applies to docstrings and comments in `crates/walrus-service/bin/client.rs`.

File: README.md [hunk 2]
```diff
@@ -92,6 +92,42 @@ curl -X PUT "http://$ADDRESS/v1/store?epochs=5" -d @"some/file" # store file `so
...
+        "config": "working_dir/client_config.yaml",
+        "command": {
+            "read": {
+                "blob_id": "4BKcDC0Ih5RJ8R0tFMz3MZVNZV8b2goT6_JiEEwNHQo"
+            }
+        }
+    }'
+```
+
+The output of a `json` command will itself be `json`-formatted, again to simplify parsing the results in a programmatic way.
```
Comment (mlegner) [inline README.md]: Suggestion: Mention piping into `jq` for manual testing.

File: crates/walrus-service/Cargo.toml [hunk 3]
```diff
@@ -31,6 +31,7 @@ regex = "1"
 reqwest = { workspace = true, features = ["json"] }
 rocksdb = "0.22.0"
 serde.workspace = true
+serde_json = { workspace = true }
```
Comment (mlegner) [inline crates/walrus-service/Cargo.toml]: Nit: We now generally use the shorthand:
```suggestion
serde_json.workspace = true
```

File: crates/walrus-service/src/client/config.rs [hunk 4]
```diff
@@ -164,9 +164,9 @@ where
...
 /// Returns the default paths for the Walrus configuration file.
 pub fn default_configuration_paths() -> Vec<PathBuf> {
-    let mut default_paths = vec!["./config.yaml".into()];
+    let mut default_paths = vec!["./client_config.yaml".into()];
     if let Some(home_dir) = home::home_dir() {
-        default_paths.push(home_dir.join(".walrus").join("config.yaml"))
+        default_paths.push(home_dir.join(".walrus").join("client_config.yaml"))
     }
     default_paths
 }
```
Comment (mlegner) [inline crates/walrus-service/src/client/config.rs]: Important: Update the docstrings in `crates/walrus-service/bin/client.rs` accordingly.

File: crates/walrus-service/bin/client.rs [hunk 5]
```diff
@@ -112,24 +134,37 @@ enum Commands {
...
         #[clap(long, action)]
+        #[serde(default)]
         dev: bool,
     },
+    /// Run the client by specifying the arguments in a json string; cli options are ignored.
+    Json {
+        /// The json-encoded command for the Walrus cli. The commands "store", "read", "publisher",
+        /// "aggregator", and "daemon", are available; "info" and "json" are not available. The json
+        /// command follows the same structure of the respective cli command. All flags are
+        /// *ignored*.
```
Comment (mlegner) [inline crates/walrus-service/bin/client.rs]: Suggestion:
```suggestion
    /// Run the client by specifying the arguments in a JSON string; CLI options are ignored.
    Json {
        /// The JSON-encoded arguments for the Walrus CLI.
        ///
        /// The JSON structure follows the CLI arguments, containing global options and a "command"
        /// object at the root level. The "command" object itself contains the command ("store",
        /// "read", "publisher", "aggregator") with an object containing the command options.
        ///
        /// Where CLI options are in "kebab-case", the respective JSON strings are in "snake_case".
        ///
        /// For example, to read a blob and write it to "some_output_file" using a specific
        /// configuration file, you can use the following JSON input:
        ///
        /// {
        ///   "config": "working_dir/client_config.yaml",
        ///   "command": {
        ///     "read": {
        ///       "blob_id": "4BKcDC0Ih5RJ8R0tFMz3MZVNZV8b2goT6_JiEEwNHQo",
        ///       "out": "some_output_file"
        ///     }
        ///   }
        /// }
        ///
        /// The commands "store", "read", "publisher", "aggregator", and "daemon", are available;
        /// "info" and "json" are not available.
        #[clap(verbatim_doc_comment)]
```

File: crates/walrus-service/bin/client.rs [hunk 6]
```diff
@@ -165,9 +214,103 @@ impl PublisherArgs {
...
+            Some(path) => {
+                write!(
+                    f,
+                    "{} Blob {} reconstructed from Walrus and written to {}.",
+                    success(),
+                    self.blob_id,
+                    path.display()
+                )
+            }
+            // The full blob has been written sto stdout.
```
Comment (mlegner) [inline crates/walrus-service/bin/client.rs]: Typo:
```suggestion
            // The full blob has been written to stdout.
```

File: crates/walrus-service/bin/client.rs [hunk 7]
```diff
@@ -165,9 +214,103 @@ impl PublisherArgs {
...
+}
+
+impl PrintOutput for StoreOutput {}
+
+/// The output of the read action.
+#[serde_as]
+#[derive(Debug, Clone, Serialize)]
+#[serde(rename_all = "lowercase")]
+struct ReadOutput {
+    out: Option<PathBuf>,
```
Comment (mlegner) [inline crates/walrus-service/bin/client.rs:290]: Suggestion: don't include this if it is `None`:
```suggestion
    #[serde(skip_serializing_if = "std::option::Option::is_none")]
    out: Option<PathBuf>,
```

File: crates/walrus-service/bin/client.rs [hunk 8]
```diff
@@ -165,9 +214,103 @@ impl PublisherArgs {
...
+    #[serde_as(as = "DisplayFromStr")]
+    blob_id: BlobId,
+    // When serializing to json, the blob is encoded as Base64 string.
```
Comment (mlegner) [inline crates/walrus-service/bin/client.rs]: Suggestion: This should probably be mentioned in the helptext for the JSON mode.

File: crates/walrus-service/bin/client.rs [hunk 9]
```diff
@@ -242,6 +374,10 @@ async fn client() -> Result<()> {
             rpc_arg: RpcArg { rpc_url },
             dev,
         } => {
+            if json {
+                // TODO: Implement the info command for json as well?
```
Comment (mlegner) [inline crates/walrus-service/bin/client.rs]: Yes, that would be great! Maybe in a follow-up PR.

File: crates/walrus-service/bin/client.rs [hunk 10]
```diff
@@ -112,24 +134,57 @@ enum Commands {
...
+        ///       "blob_id": "4BKcDC0Ih5RJ8R0tFMz3MZVNZV8b2goT6_JiEEwNHQo",
+        ///       "out": "some_output_file"
+        ///     }
+        ///   }
+        /// }
+        ///
+        /// The commands "store", "read", "publisher", "aggregator", and "daemon", are available;
+        /// "info" and "json" are not available.
+        #[clap(verbatim_doc_comment)]
+        command_string: String,
```
Comment (jpcsmith) [inline crates/walrus-service/bin/client.rs]: Suggestion: For programmatic use, stdin would be more convenient that a string, perhaps make this optional, and if none, the data is read from stdin.

File: crates/walrus-service/src/client/config.rs (hunk 4 above)
Comment (giac-mysten) [reply crates/walrus-service/src/client/config.rs]: I reverted actually. Thought it may be better to do in a PR dedicated only to that. Sorry.

Comment (mlegner) [reply crates/walrus-service/src/client/config.rs]: Agreed. 👍

File: crates/walrus-service/bin/client.rs [hunk 11]
```diff
@@ -165,9 +234,104 @@ impl PublisherArgs {
...
+/// Helper trait to get the correct string depending on the output mode.
+trait PrintOutput: Display + Serialize {
+    fn output_string(&self, json: bool) -> Result<String> {
+        if json {
+            Ok(serde_json::to_string(&self)?)
+        } else {
+            Ok(self.to_string())
+        }
+    }
+}
```
Comment (jpcsmith) [inline crates/walrus-service/bin/client.rs]: Nit: Why a trait and not just a generic function?

```rust
fn output_string<T: Display + Serialize>(value: &T, json:bool) -> Result<String>
```

File: crates/walrus-service/bin/client.rs (hunk 9 above)
Comment (giac-mysten) [reply crates/walrus-service/bin/client.rs]: Opened #465 👍

File: crates/walrus-service/bin/client.rs (hunk 11 above)
Comment (giac-mysten) [reply crates/walrus-service/bin/client.rs]: no reason, changed

File: crates/walrus-service/bin/client.rs (hunk 10 above)
Comment (giac-mysten) [reply crates/walrus-service/bin/client.rs]: very good point, changed.

File: crates/walrus-service/bin/client.rs [hunk 12]
```diff
@@ -165,9 +237,97 @@ impl PublisherArgs {
...
 async fn client() -> Result<()> {
     tracing_subscriber::fmt::init();
-    let app = App::parse();
+    let mut app = App::parse();
+    let mut json = false;
+
+    if let Commands::Json { command_string } = app.command {
+        app = serde_json::from_str(
+            &command_string.unwrap_or(std::io::read_to_string(std::io::stdin())?),
+        )?;
```
Comment (mlegner) [inline crates/walrus-service/bin/client.rs]: Suggestion: Can we add some debugging statements here? Something like this:
```suggestion
        tracing::info!("running in JSON mode");
        let command_string = match command_string {
            Some(s) => s,
            None => {
                tracing::debug!("reading JSON input from stdin");
                std::io::read_to_string(std::io::stdin())?
            }
        };
        tracing::debug!(
            command = command_string.replace("\n", ""),
            "running JSON command"
        );
        app = serde_json::from_str(&command_string)?;
```

File: crates/walrus-service/bin/client.rs [hunk 13]
```diff
@@ -184,35 +344,27 @@ async fn client() -> Result<()> {
...
+                None => {
+                    if !json {
+                        std::io::stdout().write_all(&blob)?
+                    }
                 }
-                None => std::io::stdout().write_all(&blob)?,
             }
+            println!(
+                "{}",
+                output_string(&ReadOutput { out, blob_id, blob }, json)?
```
Comment (mlegner) [inline crates/walrus-service/bin/client.rs]: Important: We probably shouldn't include the encoded blob in JSON mode if an output file is specified.

File: crates/walrus-service/bin/client.rs [hunk 14]
```diff
@@ -242,6 +394,10 @@ async fn client() -> Result<()> {
             rpc_arg: RpcArg { rpc_url },
             dev,
         } => {
+            if json {
+                // TODO: Implement the info command for JSON as well?
```
Comment (mlegner) [inline crates/walrus-service/bin/client.rs]: Nit:
```suggestion
                // TODO: Implement the info command for JSON as well. (#465)
```

File: crates/walrus-service/bin/client.rs (hunk 13 above)
Comment (giac-mysten) [reply crates/walrus-service/bin/client.rs]: good point, I pushed a simple fix in e19bf20  (not particularly elegant i'm afraid)

**Pull request merged**